File: src/shared/sealed-blob.ts
import crypto from "node:crypto";
import { type ContentAad, contentAad, openContent, sealContentWithNonce } from "./content-envelope.js";
import {
	BLOB_CHUNK_BYTES,
	BLOB_CIPHERTEXT_CHUNK_BYTES,
	BLOB_FRAME_OVERHEAD_BYTES,
	BLOB_NONCE_BYTES,
} from "./router-protocol.js";
import type { ContentEnvelope } from "./schemasContentKey.js";

export interface BlobSealContext {
	domainId: string;
	ownerSignPub: string;
	epoch: number;
	blobId: string;
}

export interface SealedBlobRange {
	bytes: Buffer;
	offset: number;
	size: number;
	epoch: number;
}

export function sealedBlobChunkCount(size: number): number {
	return Math.max(1, Math.ceil(size / BLOB_CHUNK_BYTES));
}

export function sealedBlobSize(size: number): number {
	return size + sealedBlobChunkCount(size) * BLOB_FRAME_OVERHEAD_BYTES;
}

export function blobChunkAad(context: BlobSealContext, index: number, final: boolean): ContentAad {
	return {
		domainId: context.domainId,
		ownerSignPub: context.ownerSignPub,
		epoch: context.epoch,
		kind: `blob\n${context.blobId}\n${index}\n${final ? 1 : 0}`,
	};
}

const BLOB_NONCE_TAG = "switchboard-blob-nonce-v1\n";

/**
 * Key/AAD-derived nonces match Kotlin.
 */
export function blobChunkNonce(key: Buffer, context: BlobSealContext, index: number, final: boolean): Buffer {
	return crypto
		.createHmac("sha256", key)
		.update(BLOB_NONCE_TAG)
		.update(contentAad(blobChunkAad(context, index, final)))
		.digest()
		.subarray(0, BLOB_NONCE_BYTES);
}

export function sealBlobChunk(
	plaintext: Buffer,
	key: Buffer,
	context: BlobSealContext,
	index: number,
	final: boolean,
	nonce: Buffer = blobChunkNonce(key, context, index, final),
): Buffer {
	const envelope = sealContentWithNonce(plaintext, key, blobChunkAad(context, index, final), nonce);
	return Buffer.concat([Buffer.from(envelope.nonce, "base64"), Buffer.from(envelope.ciphertext, "base64")]);
}

export function openBlobChunk(
	frame: Buffer,
	key: Buffer,
	context: BlobSealContext,
	index: number,
	final: boolean,
): Buffer {
	if (frame.length < BLOB_FRAME_OVERHEAD_BYTES) throw new Error("sealed blob frame is too short");
	const envelope: ContentEnvelope = {
		v: 1,
		epoch: context.epoch,
		nonce: frame.subarray(0, BLOB_NONCE_BYTES).toString("base64"),
		ciphertext: frame.subarray(BLOB_NONCE_BYTES).toString("base64"),
	};
	return openContent(envelope, key, blobChunkAad(context, index, final));
}

export function ciphertextRangeForPlaintext(offset: number, length: number, size: number) {
	const startIndex = Math.floor(Math.min(offset, size) / BLOB_CHUNK_BYTES);
	const end = Math.min(size, offset + length);
	const endIndex = end <= offset ? startIndex : Math.ceil(end / BLOB_CHUNK_BYTES);
	const count = Math.max(1, endIndex - startIndex);
	const ciphertextOffset = startIndex * BLOB_CIPHERTEXT_CHUNK_BYTES;
	let ciphertextLength = 0;
	for (let index = startIndex; index < startIndex + count && index < sealedBlobChunkCount(size); index++) {
		const plaintextLength = Math.max(0, Math.min(BLOB_CHUNK_BYTES, size - index * BLOB_CHUNK_BYTES));
		ciphertextLength += plaintextLength + BLOB_FRAME_OVERHEAD_BYTES;
	}
	return { startIndex, plaintextOffset: startIndex * BLOB_CHUNK_BYTES, ciphertextOffset, ciphertextLength };
}

export function openSealedBlobRange(
	range: SealedBlobRange,
	requestedOffset: number,
	requestedLength: number,
	key: Buffer,
	context: Omit<BlobSealContext, "epoch">,
): { bytes: Buffer; eof: boolean } {
	// Chunks authenticate; finalization verifies all plaintext.
	const chunks: Buffer[] = [];
	let cursor = 0;
	let index = range.offset / BLOB_CHUNK_BYTES;
	if (!Number.isInteger(index)) throw new Error("sealed blob range is not chunk-aligned");
	while (cursor < range.bytes.length) {
		const plaintextLength = Math.max(0, Math.min(BLOB_CHUNK_BYTES, range.size - index * BLOB_CHUNK_BYTES));
		const frameLength = plaintextLength + BLOB_FRAME_OVERHEAD_BYTES;
		if (cursor + frameLength > range.bytes.length) throw new Error("sealed blob range is truncated");
		const final = index + 1 === sealedBlobChunkCount(range.size);
		chunks.push(
			openBlobChunk(
				range.bytes.subarray(cursor, cursor + frameLength),
				key,
				{ ...context, epoch: range.epoch },
				index,
				final,
			),
		);
		cursor += frameLength;
		index++;
	}
	const opened = Buffer.concat(chunks);
	const trimStart = requestedOffset - range.offset;
	const bytes = opened.subarray(trimStart, trimStart + requestedLength);
	return { bytes, eof: requestedOffset + bytes.length >= range.size };
}

File: android/app/src/main/java/com/atelier_nyaarium/switchboard/crypto/SealedBlob.kt
package com.atelier_nyaarium.switchboard.crypto

import com.atelier_nyaarium.switchboard.proto.ContentEnvelope
import com.atelier_nyaarium.switchboard.proto.Protocol
import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import kotlin.math.ceil
import kotlin.math.floor

const val BLOB_NONCE_BYTES = Protocol.Wire.CONTENT_NONCE_BYTES
const val BLOB_TAG_BYTES = 16
const val BLOB_FRAME_OVERHEAD_BYTES = BLOB_NONCE_BYTES + BLOB_TAG_BYTES
const val BLOB_CIPHERTEXT_CHUNK_BYTES = Protocol.BLOB_CHUNK_BYTES + BLOB_FRAME_OVERHEAD_BYTES

data class BlobSealContext(
	val domainId: String,
	val ownerSignPub: String,
	val epoch: Int,
	val blobId: String,
)

data class SealedBlobRange(
	val bytes: ByteArray,
	val offset: Long,
	val size: Long,
	val epoch: Int,
)

data class CiphertextRange(
	val startIndex: Long,
	val plaintextOffset: Long,
	val ciphertextOffset: Long,
	val ciphertextLength: Long,
)

fun sealedBlobChunkCount(size: Long): Long = maxOf(1L, ceil(size.toDouble() / Protocol.BLOB_CHUNK_BYTES).toLong())

fun sealedBlobSize(size: Long): Long = size + sealedBlobChunkCount(size) * BLOB_FRAME_OVERHEAD_BYTES

fun blobChunkAad(context: BlobSealContext, index: Long, final: Boolean): Crypto.ContentAad =
	Crypto.ContentAad(
		context.domainId,
		context.ownerSignPub,
		context.epoch,
		"blob\n${context.blobId}\n$index\n${if (final) 1 else 0}",
	)

private val BLOB_NONCE_TAG = "switchboard-blob-nonce-v1\n".toByteArray(Charsets.UTF_8)

/** Matches TypeScript nonce. */
fun blobChunkNonce(key: ByteArray, context: BlobSealContext, index: Long, final: Boolean): ByteArray {
	val mac = Mac.getInstance("HmacSHA256")
	mac.init(SecretKeySpec(key, "HmacSHA256"))
	mac.update(BLOB_NONCE_TAG)
	mac.update(blobChunkAad(context, index, final).bytes())
	return mac.doFinal().copyOfRange(0, BLOB_NONCE_BYTES)
}

/** One sealed frame. */
fun sealBlobChunk(
	plaintext: ByteArray,
	key: ByteArray,
	context: BlobSealContext,
	index: Long,
	final: Boolean,
	nonce: ByteArray = blobChunkNonce(key, context, index, final),
): ByteArray {
	val envelope = Crypto.sealContent(plaintext, key, blobChunkAad(context, index, final), nonce)
	return Base64.getDecoder().decode(envelope.nonce) + Base64.getDecoder().decode(envelope.ciphertext)
}

fun openBlobChunk(
	frame: ByteArray,
	key: ByteArray,
	context: BlobSealContext,
	index: Long,
	final: Boolean,
): ByteArray {
	if (frame.size < BLOB_FRAME_OVERHEAD_BYTES) throw IllegalArgumentException("sealed blob frame is too short")
	val envelope = ContentEnvelope(
		v = 1L,
		epoch = context.epoch.toLong(),
		nonce = Base64.getEncoder().encodeToString(frame.copyOfRange(0, BLOB_NONCE_BYTES)),
		ciphertext = Base64.getEncoder().encodeToString(frame.copyOfRange(BLOB_NONCE_BYTES, frame.size)),
	)
	return Crypto.openContent(envelope, key, blobChunkAad(context, index, final))
}

fun ciphertextRangeForPlaintext(offset: Long, length: Long, size: Long): CiphertextRange {
	val startIndex = floor(minOf(offset, size).toDouble() / Protocol.BLOB_CHUNK_BYTES).toLong()
	val end = minOf(size, offset + length)
	val endIndex = if (end <= offset) startIndex else ceil(end.toDouble() / Protocol.BLOB_CHUNK_BYTES).toLong()
	val count = maxOf(1L, endIndex - startIndex)
	var ciphertextLength = 0L
	for (index in startIndex until (startIndex + count)) {
		if (index >= sealedBlobChunkCount(size)) break
		val plaintextLength = maxOf(0L, minOf(Protocol.BLOB_CHUNK_BYTES.toLong(), size - index * Protocol.BLOB_CHUNK_BYTES))
		ciphertextLength += plaintextLength + BLOB_FRAME_OVERHEAD_BYTES
	}
	return CiphertextRange(
		startIndex = startIndex,
		plaintextOffset = startIndex * Protocol.BLOB_CHUNK_BYTES,
		ciphertextOffset = startIndex * BLOB_CIPHERTEXT_CHUNK_BYTES,
		ciphertextLength = ciphertextLength,
	)
}

fun openSealedBlobRange(
	range: SealedBlobRange,
	requestedOffset: Long,
	requestedLength: Long,
	key: ByteArray,
	context: BlobSealContext,
): Pair<ByteArray, Boolean> {
	if (range.offset % Protocol.BLOB_CHUNK_BYTES != 0L) throw IllegalArgumentException("sealed blob range is not chunk-aligned")
	val chunks = ArrayList<ByteArray>()
	var cursor = 0
	var index = range.offset / Protocol.BLOB_CHUNK_BYTES
	while (cursor < range.bytes.size) {
		val plaintextLength = maxOf(0L, minOf(Protocol.BLOB_CHUNK_BYTES.toLong(), range.size - index * Protocol.BLOB_CHUNK_BYTES))
		// Fixed stride. Final chunk may be short.
		val frameLength = plaintextLength + BLOB_FRAME_OVERHEAD_BYTES
		if (frameLength > Int.MAX_VALUE || cursor.toLong() + frameLength > range.bytes.size) {
			throw IllegalArgumentException("sealed blob range is truncated")
		}
		val final = index + 1L == sealedBlobChunkCount(range.size)
		chunks += openBlobChunk(
			range.bytes.copyOfRange(cursor, cursor + frameLength.toInt()),
			key,
			context.copy(epoch = range.epoch),
			index,
			final,
		)
		cursor += frameLength.toInt()
		index++
	}
	val opened = ByteArray(chunks.sumOf { it.size })
	var destination = 0
	for (chunk in chunks) {
		chunk.copyInto(opened, destination)
		destination += chunk.size
	}
	val trimStart = requestedOffset - range.offset
	val start = trimStart.coerceIn(0L, opened.size.toLong()).toInt()
	val end = (trimStart + requestedLength).coerceIn(start.toLong(), opened.size.toLong()).toInt()
	val bytes = opened.copyOfRange(start, end)
	return bytes to (requestedOffset + bytes.size >= range.size)
}

fun openSealedBlobRange(
	bytes: ByteArray,
	offset: Long,
	size: Long,
	epoch: Int,
	requestedOffset: Long,
	requestedLength: Long,
	key: ByteArray,
	domainId: String,
	ownerSignPub: String,
	blobId: String,
): Pair<ByteArray, Boolean> = openSealedBlobRange(
	SealedBlobRange(bytes, offset, size, epoch),
	requestedOffset,
	requestedLength,
	key,
	BlobSealContext(domainId, ownerSignPub, epoch, blobId),
)
